src/System.IO.Compression/src/System/IO/Compression/DeflaterZLib.cs

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Diagnostics.Contracts;
+using System.Runtime.InteropServices;
+using System.Security;
+
+using ZErrorCode = System.IO.Compression.ZLibNative.ErrorCode;
+using ZFlushCode = System.IO.Compression.ZLibNative.FlushCode;
+
+namespace System.IO.Compression
+{
+    internal class DeflaterZLib : IDeflater
+    {
+        private ZLibNative.ZLibStreamHandle _zlibStream;
+        private GCHandle? _inputBufferHandle;
+        private bool _isDisposed;
+
+        // Note, DeflateStream or the deflater do not try to be thread safe.
+        // The lock is just used to make writing to unmanaged structures atomic to make sure
+        // that they do not get inconsistent fields that may lead to an unmanaged memory violation.
+        // To prevent *managed* buffer corruption or other weird behaviour users need to synchronise
+        // on the stream explicitly.
+        private Object syncLock = new Object();
```

Comment:
It should also be `object` not `Object` (use language keywords instead of BCL types).